tests/unit/test_config.py

```diff
@@ -0,0 +1,133 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""Test functionality in the config module."""
+import os
+import stat
+
+import configargparse
+
+import snafu.config
+
+
+def test_check_file_returns_bool_as_expected(tmpdir):
+    """Test that the check_file function will return ``True`` and ``False`` when it is supposed to."""
+
+    tmpdir.chmod(stat.S_IREAD | stat.S_IWRITE | stat.S_IEXEC)
+    test_file = tmpdir.join("testfile.txt")
+    test_file.write("some content")
+    test_file_path = test_file.realpath()
+    assert test_file.check()
+
+    test_perms = (
+        (os.R_OK, stat.S_IREAD),
+        (os.R_OK | os.W_OK, stat.S_IREAD | stat.S_IWRITE),
+        (os.R_OK | os.W_OK | os.EX_OK, stat.S_IREAD | stat.S_IWRITE | stat.S_IEXEC),
+    )
+
+    with tmpdir.as_cwd():
+        for check, perm in test_perms:
+            test_file.chmod(0)
+            assert snafu.config.check_file(test_file_path, perms=check) is False
+            test_file.chmod(perm)
+            assert snafu.config.check_file(test_file_path, perms=check) is True
+
+
+def test_none_or_type_function():
+    """Test that the none_or_type function returns function that casts values only when they aren't None."""
+
+    tests = (
+        (int, 1),
+        (int, "1"),
+        (str, "hey"),
+        (str, 2),
+        (dict, (("a", 1), ("b", 2))),
+    )
+
+    for expected_type, value in tests:
+        wrapped = snafu.config.none_or_type(expected_type)
+        assert wrapped(None) is None
+        assert wrapped(value) == expected_type(value)
+        assert isinstance(wrapped(value), expected_type)
+
+
+class TestConfig:
+    """Test functionality of Config class."""
+
+    test_args = (
+        snafu.config.ConfigArgument("param1", type=int),
+        snafu.config.ConfigArgument("param2", type=str),
+        snafu.config.ConfigArgument("--param3", type=str, env_var="param3"),
+        snafu.config.ConfigArgument("--param4", type=int, env_var="p4"),
+    )
+    test_input = ["1234", "my_string", "--param3", "1234", "--param4", "4321"]
+
+    @staticmethod
+    def verify_args(config: snafu.config.Config):
+        """Verify that config contains expected args from cls.test_args and cls.test_input."""
+
+        assert config.param1 == 1234
+        assert config.param2 == "my_string"
+        assert config.param3 == "1234"
+        assert config.param4 == 4321
+
+    @staticmethod
+    def get_config_instance() -> snafu.config.Config:
+        """Return a new Config instance with tool_name set to 'TEST'."""
+        try:
+            configargparse.init_argument_parser()
+        except ValueError:
+            del configargparse._parsers["default"]  # pylint: disable=W0212
+            configargparse.init_argument_parser()
```

Comment:
Why do we need these lines?

Reply:
[configargparse](https://pypi.org/project/ConfigArgParse/) uses a global argument parser which we are taking advantage of, but in order to have a clean parser for each test we need to make sure that we refresh it, and that's what these lines do. Here are the relevant lines in configargparse: https://github.com/bw2/ConfigArgParse/blob/master/configargparse.py#L35